File: android/app/src/main/kotlin/org/fgsec/halogen/components/LoadErrorView.kt
package org.fgsec.halogen.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch

/// Full-screen load-failure state with a working Retry. Error branches render
/// this instead of a bare empty state: that branch has no scroll container,
/// so pull-to-refresh never works there — without the button the only
/// recovery was leaving the screen.
@Composable
fun LoadErrorView(
    title: String,
    message: String,
    retry: suspend () -> Unit,
) {
    var retrying by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    Column(
        Modifier.fillMaxSize().padding(32.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp, Alignment.CenterVertically),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Icon(
            halogenIcon("wifi.exclamationmark"),
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.size(48.dp),
        )
        Text(title, style = MaterialTheme.typography.titleMedium)
        Text(
            message,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
        )
        Button(
            onClick = {
                if (retrying) return@Button
                retrying = true
                scope.launch {
                    retry()
                    retrying = false
                }
            },
            modifier = Modifier.testTag("load-error-retry"),
        ) {
            if (retrying) {
                CircularProgressIndicator(
                    modifier = Modifier.size(16.dp),
                    color = MaterialTheme.colorScheme.onPrimary,
                    strokeWidth = 2.dp,
                )
            } else {
                Text("Retry")
            }
        }
    }
}

File: ios/Halogen/Components/LoadErrorView.swift
import SwiftUI

/// Full-screen load-failure state with a working Retry. Error branches render
/// this instead of a bare ContentUnavailableView: that branch has no scroll
/// container, so pull-to-refresh never works there — without the button the
/// only recovery was leaving the screen.
struct LoadErrorView: View {
    let title: String
    let message: String
    let retry: () async -> Void

    @State private var retrying = false

    var body: some View {
        ContentUnavailableView {
            Label(title, systemImage: "wifi.exclamationmark")
        } description: {
            Text(message)
        } actions: {
            Button {
                guard !retrying else { return }
                retrying = true
                Task {
                    await retry()
                    retrying = false
                }
            } label: {
                if retrying {
                    ProgressView().controlSize(.small)
                } else {
                    Text("Retry")
                }
            }
            .buttonStyle(.borderedProminent)
            .accessibilityIdentifier("load-error-retry")
        }
    }
}
